File: src/main/java/Library/Book.java
package Library;

import java.time.LocalDate;
import java.util.Optional;

// Book: Handing the book's parameters
public class Book {


    private final String title;
    private final String author;

    public enum Status { Available, Checked_out, On_hold }
    private Status status = Status.Available;

    private Optional<LocalDate> dueDate = Optional.empty();


    // Book initialized
    public Book(String title, String author){
        this.title = title;
        this.author = author;
    }

    // getters
    public String getTitle(){
        return title;
    }
    public String getAuthor(){
        return author;
    }
    public Status getStatus(){
        return status;
    }
    public Optional<LocalDate> getDueDate() {
        return dueDate;
    }
    // Output the due date
    public String getDueDateText(){
        return dueDate.map(LocalDate::toString).orElse("-");
    }

    // setters
    public void setStatus(Status status){
        this.status = status;
        if (status != Status.Checked_out) {
            dueDate = Optional.empty();
        }
    }
    public void setDueDate(LocalDate date){
        this.dueDate = Optional.ofNullable(date);
    }
}

File: A3/models/Book.js
// Book.js

// Enum should NOT be recreated for every instance
const Status = {
    Available: "Available",
    Checked_out: "Checked_out",
    On_hold: "On_hold"
};

class Book {
    constructor(title, author) {
        this.title = title;
        this.author = author;

        this.status = Status.Available;
        this.dueDate = null;
    }

    // getters
    getTitle() { return this.title; }
    getAuthor() { return this.author; }
    getStatus() { return this.status; }
    getDueDate() { return this.dueDate; }

    getDueDateText() {
        return this.dueDate ? this.dueDate.toISOString().split("T")[0] : "-";
    }

    // setters
    setStatus(status) {
        this.status = status;
        if (status !== Status.Checked_out) {
            this.dueDate = null;
        }
    }

    setDueDate(date) {
        this.dueDate = date ? new Date(date) : null;
    }
}

// expose the enum
Book.Status = Status;

module.exports = Book;
